woocommerce/woocommerce-gateway-paypal-express-checkout
# Add support for PayPal JS SDK

### Description

**Work in progress.** Intended to supersede #622.

This PR adds support for the new [JS SDK integration](https://developer.paypal.com/docs/checkout/) for Smart Payment Buttons.
It doesn't directly replace the current implementation (which uses [`checkout.js`](https://developer.paypal.com/docs/archive/checkout/)). Instead, depending on whether the REST API credentials are available or not, it uses either the JS SDK or the old implementation.

The code also tries to appropriately map properties/functionality from the [old implementation to the new one](https://developer.paypal.com/docs/checkout/reference/upgrade-integration).

### To get your Client ID and secret
1. Go to https://developer.paypal.com/developer/applications/ and log into your sandbox developer account
2. Go to **My Apps & Credentials** from the nav menu.
3. Under **REST API apps**, click **Create App**.
4. Give it a name and drop the dropdown select the merchant sandbox credentials you use on your site. 
5. Copy the Client ID
6. Click _Show_ under **Secret** and copy it. 

Docs: https://developer.paypal.com/docs/checkout/integrate/#1-set-up-your-development-environment

### Steps to test
1. Get the REST API credentials (Client ID and secret) from your PayPal dashboard.
2. Go to the plugin settings:
2.1. Enter the live/sandbox REST API credentials.
2.2. Make sure the "Smart Payment Buttons" setting is enabled.
2.3. For testing purposes, enable PayPal Checkout for the single product page and cart page.
3. Make sure the PayPal button appears as configured, is clickable and completes the payment process for each of these scenarios:

    <details>
    <summary>Test Case: Regular checkout process</summary>
      1. Add an item to the cart.<br />
      2. Go to the cart page.<br />
      3. Go to the checkout page.<br />
      4. Use the PayPal button to complete the payment process.
   </details>

    <details>
    <summary>Test Case: Pay directly from a product's page</summary>
      1. Open an item's product page.<br />
      2. Use the PayPal button to complete the payment process.
   </details>

    <details>
    <summary>Test Case: Pay directly from the cart page</summary>
      1. Add an item to the cart.<br />
      2. Go to the cart page.
      3. Use the PayPal button to complete the payment process.
   </details>

Closes #626.
